tutorials/iot/graphical-snaps/graphical-snaps.md

```diff
@@ -0,0 +1,608 @@
+We will revisit this later to ensure the snap can be compiled to function on other architectures.
+
+
+## First Pass Snapping: The Final Bits
+duration: 4:00
+
+We need to set those two environment variables before executing the binary inside the snap. One option is a simple launching script:
+```
+#!/bin/bash
+LIBGL_DRIVERS_PATH=$SNAP/usr/lib/x86_64-linux-gnu/dri \
+XDG_RUNTIME_DIR=$XDG_RUNTIME_DIR/.. \    
+$SNAP/usr/bin/glmark2-wayland
+```
+
+and point the “command:” in the snapcraft.yaml file to it (don’t forget to install it in the snap!!)
+
+Another option (which we will use here) is to adjust the `command:` file like this:
+```yaml
+...
+    command: "bash -c 'XDG_RUNTIME_DIR=$(dirname $XDG_RUNTIME_DIR) LIBGL_DRIVERS_PATH=$(find $SNAP/usr/lib/ -name dri) $SNAP/usr/bin/glmark2-wayland'"
+...
+```
+
+### The working YAML file:
+
+```yaml
+name: glmark2-wayland
+version: 0.1
+summary: GLMark2 on Wayland
+description: |
+  GLMark2 on Wayland
+confinement: devmode
+grade: devel
+
+apps:
+  glmark2-wayland:
+    command: "bash -c 'XDG_RUNTIME_DIR=$(dirname $XDG_RUNTIME_DIR) LIBGL_DRIVERS_PATH=$(find $SNAP/usr/lib/ -name dri) $SNAP/usr/bin/glmark2-wayland'"
+    plugs:
+      - opengl
+      - wayland
+
+parts:
+  glmark2-wayland:
+    plugin: nil
+    stage-packages:
+      - glmark2-wayland
+
+passthrough:
+  layout:
+    /usr/share/glmark2:
+      bind: $SNAP/usr/share/glmark2
+```
+
+Rebuild the snap and install it - “snap run glmark2-wayland” should work fine.
+
+```bash
+snapcraft cleanbuild
+sudo snap install --dangerous ./glmark2-wayland_0.1_amd64.snap --devmode
+miral-kiosk&
+snap run glmark2-wayland
+```
+
+Inside the Mir-on-X window, you should see the same graphical animations you saw earlier, and statistics printed to your console. The difference is that this time they are coming from a snap.
+
+## First Pass Snapping: Notes
+duration: 1:00
+
+### Why did we step through that process instead of simply presenting the working .yaml?
+
+Because snapping applications can reveal lots of hard-coded paths and assumptions that applications make, which snap confinement will break. It is good to understand the steps needed to debug and solve these problems.
+
+There can be many, many environment variables and support files that need to be set up inside snaps, for applications to run correctly. Much of this work has already been done and automated in the *snapcraft-desktop-helpers* project, we will be using this in a follow-up tutorial.
+
+### We now have a snap working with a desktop Wayland server
+ 
+*But our goal is to have it working on Ubuntu Core on your chosen device.*
+
+## Second Pass Snapping: Ubuntu Core
+duration: 5:00
+
+### Ubuntu Core Setup
+
+Once you have set up Ubuntu Core on your device and logged in install the “mir-kiosk” snap.
+
+```bash
+snap install --beta mir-kiosk
+```
+
+Now you should have a black screen with a white mouse cursor.
```

Comment:
Does the mouse cursor show up in qemu VM, too? I don't see it (but black screen, yes).

Reply:
I don't have a qemu VM with gl support. @gerboland wrote notes saying the white cursor appears and I see it on dragonboard. If it is not visible something is not working right.